site/public/js/testcase-output.js

```diff
@@ -82,22 +173,62 @@ function loadTestcaseOutput(div_name, gradeable_id, who_id, index, version = '')
                 if (loading_testcases_xml_http_requests.has(orig_div_name) && loading_testcases_xml_http_requests.get(orig_div_name).readyState === 0) {
                     loading_testcases_xml_http_requests.delete(orig_div_name);
-                    CollapseTestcaseOutput(orig_div_name, index, loadingTools);
-                    console.log('JSON Load Aborted');
+                    CollapseTestcaseOutput(orig_div_name, index, num_test_cases, loadingTools);
                 }
                 else {
-                    alert('Could not load diff, please refresh the page and try again.');
-                    console.log(e);
-                    // eslint-disable-next-line no-undef
-                    displayAjaxError(e);
+                    // Display error message once
+                    loading_testcases_xml_http_requests.delete(orig_div_name);
+                    if (loading_testcases_xml_http_requests.size === 0) {
+                        alert('Could not load diff, please refresh the page and try again.');
+                        console.log(e);
+                        // eslint-disable-next-line no-undef
+                        displayAjaxError(e);
+                    }
                 }
             },
         }));
     }
 }
+
+/**
+ * Expands or Collapses all test case outputs.
+ *
+ * @global {loading_testcases_xml_http_requests} : Map (Key - string [orig_div_name]; Value - XMLHttpRequest object)
+ *                        to store the XMLHttpRequest object returned by getJSON when getting the output to display
+ *
+ * @param {string} div_name - ID of div that calls this function
+ * @param {boolean} expand_all - should all test cases be expanded
+ * @param {int} num_test_cases - number of test cases
+ * @param {string} gradeable_id - ID of the gradeable
+ * @param {string} who_id - ID of the submitter
+ * @param {number} version - submission version
+ */
+// eslint-disable-next-line no-unused-vars
+function loadAllTestcaseOutput(div_name, expand_all, num_test_cases, gradeable_id, who_id, version = '') {
+    // Check if toggle should not be interacted with
+    if ($(`#${div_name}`).css('cursor') === 'default') {
+        return;
+    }
+
+    // Expand/Collapse all test cases
+    for (let index = 0; index < num_test_cases; index++) {
+        // Check if test case has loading tools (can be expanded/collapsed)
+        if ($(`#tc_${index}`).find('.loading-tools').length === 1) {
+            const test_case_div_name = `testcase_${index}`;
+            const isTestCaseLoadingToolsShowVisible = $(`#tc_${index}`).find('.loading-tools').find('.loading-tools-show').is(':visible');
+
+            // Check if test case should be expanded/collapsed
+            if ((expand_all && isTestCaseLoadingToolsShowVisible) || (!expand_all && !isTestCaseLoadingToolsShowVisible)) {
+                loadTestcaseOutput(test_case_div_name, gradeable_id, who_id, index, num_test_cases, version);
```

Comment:
Execution comes to this function when `ExpandAll` or `CollapseAll` is clicked. Through this function `CheckStatesOfAllTestcases()` function is called for each test case and that tests state of each test case. Is that needed because you know after ExpandAll/CollapseAll is clicked state of the all the test cases is going to be same. You could pass some sort of a boolean to stop that function from executing.